### ISSUE TITLE
Chords need a way for executing a callback even if a task in the task group fails

### ISSUE THREAD

**Issue by jdelic** [celery/celery#1881]
Pre Celery-3.1, the callback (or body) of a `chord` was executed always. This is no longer the case. Instead, if at least one task of the chord's task group throws an exception, the exception is propagated to the `chord` instance and its `status` attribute is set to `FAILURE`.

Now take the following example: A chord is used to contact 400 network servers to query their status in parallel. Once all 400 servers have been contacted, a report should be compiled about which servers were available and which weren't. Temporary outages are expected.

Pre-3.1 the chord's callback could iterate over its `parent` attribute and easily compile that report. Post-3.1 there seems to be no way to execute a task after all the chord's tasks have been executed by a worker at least once, even though they might have failed. I experimented with `link_error` and signals, but nothing seems to fill this need right now.

I therefor propose to either add an additional parameter to `chord` (`link_error` would probably be intuitive) which gets called if an error had occurred during task processing or instead call the callback regardless.

**Comment by alejandrodnm** [comment]
Hi @ask I have the same use case as @jdelic, I need to execute a group of tasks that are very volatile (Texture and animation exportations for mobile games, calling Adobe Flash, Texture Packer, Unity, etc) and I need to make a report at the end that contains the result and status of every task, so I'm going to use a chord.

For what I see in your commit, and correct me if I'm wrong, when a task fails in a chord you execute a new group of tasks that are specified in the `link_error` options of the chord callback.

So we can do this `chord(tasks)(callback.s().set(link_error=['error_callback']))` and when an exception is raise in one of the tasks the `error_callback` will execute getting as an argument, what I think it's the id of the chord.

I'm trying this for an `error_callback`:

```python
@task(bind=True)
def error_callback(self, *args, **kwargs):
    chord_task = self.app.AsyncResult(args[0])    
    children = chord_task.children
```
`children` ends up with a GroupResult that only contains the AsyncResult corresponding to `error_callback`. 

What I'm trying to ask is there any way to get a list of the tasks that belong to the chord in the error_callback, so I can then loop through their status?

I'm new in this project and the team tells me that our tasks catch all the exceptions and have a unified return policy for errors, the task will not raise exceptions but will return the error. In theory the error callback will never be called and I'll get all the info I need in the success callback, but I just want a backup for unexpected events.

**Comment by maccinza** [comment]
Hello guys, what was the resolution for this?
I have the exact same need expressed by alejandrodnm and jdelic: I would like to get the successful results of a chord for which some dependencies might be failing.

**Comment by alejandrodnm** [comment]
This was a long time ago, I wrote this answer I don´t know if it helps

https://stackoverflow.com/questions/29355613/run-a-chord-callback-even-if-the-main-tasks-fail/31454228#31454228

**Comment by maccinza** [comment]
Thanks for the feedback alejandrodnm.
I've read that answer before but I was trying to avoid using a datastore for keeping the results of all the tasks and fetch it later.
I might be really wrong about this but it makes no sense to me that the scenario exposed in this issue is not achievable. The whole chord failing because of a single failing task, giving you no chance to collect information about the successful ones at a synchronisation point is so frustrating.

**Comment by alejandrodnm** [comment]
@maccinza sorry that's all I remember I've haven't worked with celery or python in a long time. Cheers

**Comment by maccinza** [comment]
@alejandrodnm 
Thanks for the help and feedback mate. Cheers!

**Comment by jdelic** [comment]
after I wrote this issue I had passed the job off to another developer and I don't remember how we proceeded with it :(

However, I totally agree that if the scenario that I first described isn't achievable without an additional datastore that's a bug. Especially since it definitely was possible pre-3.1.

@ask I apologize for pinging you on this closed issue, but I want to make you aware of it... should they file a new issue?

**Comment by thedrow** [comment]
Has any of you tried calling `chord_result.parent`?
Does that provide you with the group linked to the chord?

**Comment by ziXet** [comment]
Monkey patching could be a temporary solution for this issue.
In case of Redis backend:
`
def _unpack_chord_result(self, tup, decode,
                         EXCEPTION_STATES=states.EXCEPTION_STATES,
                         PROPAGATE_STATES=states.PROPAGATE_STATES):
    _, tid, state, retval = decode(tup)
    if state in EXCEPTION_STATES:
        retval = self.exception_to_python(retval)
    if state in PROPAGATE_STATES:
        return None
        # raise ChordError('Dependency {0} raised {1!r}'.format(tid, retval))
    return retval

RedisBackend._unpack_chord_result = _unpack_chord_result
`

**Comment by TariqAbughofa** [comment]
For those using a database as a backend and want to be able to do this. You have to add this task to override the builtin `chord_unlock`:

```
@app.task(name='celery.chord_unlock', max_retries=None, shared=False,
          default_retry_delay=app.conf.result_chord_retry_interval, ignore_result=True, lazy=False, bind=True)
def unlock_chord(self, group_id, callback, interval=None,
                 max_retries=None, result=None,
                 Result=app.AsyncResult, GroupResult=app.GroupResult,
                 result_from_tuple=result_from_tuple, **kwargs):
    if interval is None:
        interval = self.default_retry_delay

    # check if the task group is ready, and if so apply the callback.
    callback = maybe_signature(callback, app)
    deps = GroupResult(
        group_id,
        [result_from_tuple(r, app=app) for r in result],
        app=app,
    )
    j = deps.join_native if deps.supports_native_join else deps.join

    try:
        ready = deps.ready()
    except Exception as exc:
        raise self.retry(
            exc=exc, countdown=interval, max_retries=max_retries,
        )
    else:
        if not ready:
            raise self.retry(countdown=interval, max_retries=max_retries)

    callback = maybe_signature(callback, app=app)
    try:
        with allow_join_result():
            ret = [_ for _ in j(
                timeout=app.conf.result_chord_join_timeout,
                propagate=False,
            ) if not isinstance(_, Exception)]
    except Exception as exc:  # pylint: disable=broad-except
        try:
            culprit = next(deps._failed_join_report())
            reason = f'Dependency {culprit.id} raised {exc!r}'
        except StopIteration:
            reason = repr(exc)
        logger.exception('Chord %r raised: %r', group_id, exc)
        app.backend.chord_error_from_stack(callback, ChordError(reason))
    else:
        try:
            callback.delay(ret)
        except Exception as exc:  # pylint: disable=broad-except
            logger.exception('Chord %r raised: %r', group_id, exc)
            app.backend.chord_error_from_stack(
                callback,
                exc=ChordError(f'Callback error: {exc!r}'),
            )
```
this is an exact copy of: `celery.app.builtins.add_unlock_chord_task.unlock_chord` except:
```
            ret = j(
                timeout=app.conf.result_chord_join_timeout,
                propagate=True,
            )
``` 
is replaced by:
```
            ret = [_ for _ in j(
                timeout=app.conf.result_chord_join_timeout,
                propagate=False,
            ) if isinstance(_, Exception)]
```

* tested on celery 5.2.3

**Comment by gautierrog** [comment]
> For those using a database as a backend and want to be able to do this. You have to add this task to override the builtin `chord_unlock`:
> 
> ```
> @app.task(name='celery.chord_unlock', max_retries=None, shared=False,
>           default_retry_delay=app.conf.result_chord_retry_interval, ignore_result=True, lazy=False, bind=True)
> def unlock_chord(self, group_id, callback, interval=None,
>                  max_retries=None, result=None,
>                  Result=app.AsyncResult, GroupResult=app.GroupResult,
>                  result_from_tuple=result_from_tuple, **kwargs):
>     if interval is None:
>         interval = self.default_retry_delay
> 
>     # check if the task group is ready, and if so apply the callback.
>     callback = maybe_signature(callback, app)
>     deps = GroupResult(
>         group_id,
>         [result_from_tuple(r, app=app) for r in result],
>         app=app,
>     )
>     j = deps.join_native if deps.supports_native_join else deps.join
> 
>     try:
>         ready = deps.ready()
>     except Exception as exc:
>         raise self.retry(
>             exc=exc, countdown=interval, max_retries=max_retries,
>         )
>     else:
>         if not ready:
>             raise self.retry(countdown=interval, max_retries=max_retries)
> 
>     callback = maybe_signature(callback, app=app)
>     try:
>         with allow_join_result():
>             ret = [_ for _ in j(
>                 timeout=app.conf.result_chord_join_timeout,
>                 propagate=False,
>             ) if not isinstance(_, Exception)]
>     except Exception as exc:  # pylint: disable=broad-except
>         try:
>             culprit = next(deps._failed_join_report())
>             reason = f'Dependency {culprit.id} raised {exc!r}'
>         except StopIteration:
>             reason = repr(exc)
>         logger.exception('Chord %r raised: %r', group_id, exc)
>         app.backend.chord_error_from_stack(callback, ChordError(reason))
>     else:
>         try:
>             callback.delay(ret)
>         except Exception as exc:  # pylint: disable=broad-except
>             logger.exception('Chord %r raised: %r', group_id, exc)
>             app.backend.chord_error_from_stack(
>                 callback,
>                 exc=ChordError(f'Callback error: {exc!r}'),
>             )
> ```
> 
> this is an exact copy of: `celery.app.builtins.add_unlock_chord_task.unlock_chord` except:
> 
> ```
>             ret = j(
>                 timeout=app.conf.result_chord_join_timeout,
>                 propagate=True,
>             )
> ```
> 
> is replaced by:
> 
> ```
>             ret = [_ for _ in j(
>                 timeout=app.conf.result_chord_join_timeout,
>                 propagate=False,
>             ) if not isinstance(_, Exception)]
> ```
> 
> * tested on celery 5.2.3

It works. Would it deserve to be in a PR?

**Comment by swssl** [comment]
> * tested on celery 5.2.3

Thanks @TariqAbughofa  and @gautierrog . If anybody needs this in the future, this one works on my machine with version 5.6.2: 
```python
from celery.backends.base import _create_chord_error_with_cause
from celery.canvas import maybe_signature
from celery.result import allow_join_result, result_from_tuple
from celery.utils.log import get_task_logger


@app.task(
    name="celery.chord_unlock",
    max_retries=None,
    shared=False,
    default_retry_delay=app.conf.result_chord_retry_interval,
    ignore_result=True,
    lazy=False,
    bind=True,
)
def unlock_chord(
    self,
    group_id,
    callback,
    interval=None,
    max_retries=None,
    result=None,
    Result=app.AsyncResult,
    GroupResult=app.GroupResult,
    result_from_tuple=result_from_tuple,
    **kwargs,
):
    logger = get_task_logger(__name__)
    if interval is None:
        interval = self.default_retry_delay

    # check if the task group is ready, and if so apply the callback.
    callback = maybe_signature(callback, app)
    deps = GroupResult(
        group_id,
        [result_from_tuple(r, app=app) for r in result],
        app=app,
    )
    j = deps.join_native if deps.supports_native_join else deps.join

    try:
        ready = deps.ready()
    except Exception as exc:
        raise self.retry(
            exc=exc,
            countdown=interval,
            max_retries=max_retries,
        )
    else:
        if not ready:
            raise self.retry(countdown=interval, max_retries=max_retries)

    callback = maybe_signature(callback, app=app)
    try:
        with allow_join_result():
            ret = [
                _
                for _ in j(
                    timeout=app.conf.result_chord_join_timeout,
                    propagate=False,
                )
                if not isinstance(_, Exception)
            ]
    except Exception as exc:  # pylint: disable=broad-except
        try:
            culprit = next(deps._failed_join_report())
            reason = f"Dependency {culprit.id} raised {exc!r}"
        except StopIteration:
            reason = repr(exc)
        logger.exception("Chord %r raised: %r", group_id, exc)
        chord_error = _create_chord_error_with_cause(message=reason, original_exc=exc)
        app.backend.chord_error_from_stack(callback=callback, exc=chord_error)
    else:
        try:
            callback.delay(ret)
        except Exception as exc:  # pylint: disable=broad-except
            logger.exception("Chord %r raised: %r", group_id, exc)
            chord_error = _create_chord_error_with_cause(
                message=f"Callback error: {exc!r}", original_exc=exc
            )
            app.backend.chord_error_from_stack(callback=callback, exc=chord_error)
```
Still, changes are only necessary in the `ret = ...` part but the code around it has changed since.

**Comment by auvipy** [comment]
contributions are welcome
